cmd/fabric8-jenkins-idler/idler.go

```diff
@@ -24,123 +25,138 @@ const (
+	idlerLogger.Info("Starting all Idler workers")
+
+	// Create synchronized map for UserIdler instances
+	userIdlers := openshift.NewUserIdlerMap()
+
+	// Start the controllers to monitor the OpenShift clusters
+	idler.startOpenShiftControllers(ctx, wg, cancel, userIdlers)
 
 	// Start API router
 	go func() {
 		// Create and start a Router instance to serve the REST API
-		idlerAPI := api.NewIdlerAPI(openShift, controller)
+		idlerAPI := api.NewIdlerAPI(userIdlers, idler.clusterView)
 		router := router.NewRouter(router.CreateAPIRouter(idlerAPI))
 
 		router.Start(ctx, wg, cancel)
 	}()
 
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		go func() {
-			mainLogger.Info("Starting to watch OpenShift deployment configuration changes.")
-			if err := openShift.WatchDeploymentConfigs("", "-jenkins", controller.HandleDeploymentConfig); err != nil {
-				cancel()
-				return
-			}
-		}()
-
-		for {
-			select {
-			case <-ctx.Done():
-				mainLogger.Infof("Stopping to watch OpenShift deployment configuration changes.")
-				cancel()
-				return
-			}
-		}
-	}()
-
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		go func() {
-			mainLogger.Info("Starting to watch OpenShift build configuration changes.")
-			if err := openShift.WatchBuilds("", "JenkinsPipeline", controller.HandleBuild); err != nil {
-				cancel()
-				return
-			}
-		}()
-
-		for {
-			select {
-			case <-ctx.Done():
-				mainLogger.Infof("Stopping to watch OpenShift build configuration changes.")
-				cancel()
-				return
-			}
-		}
-	}()
-
 	if addProfiler {
 		go func() {
-			mainLogger.Infof("Starting profiler on port %d", profilerPort)
+			idlerLogger.Infof("Starting profiler on port %d", profilerPort)
 			router := router.NewRouterWithPort(httprouter.New(), profilerPort)
 			router.Start(ctx, wg, cancel)
 		}()
 	}
 }
 
+func (idler *Idler) startOpenShiftControllers(ctx context.Context, wg *sync.WaitGroup, cancel context.CancelFunc, userIdlers *openshift.UserIdlerMap) {
+	openShiftClient := client.NewOpenShift()
+
+	for _, openShiftCluster := range idler.clusterView.GetClusters() {
+		// Create Controller
+		controller := openshift.NewController(
+			ctx,
+			openShiftCluster.APIURL,
+			openShiftCluster.Token,
+			userIdlers,
+			idler.tenantService,
+			idler.featureService,
+			idler.config,
+			wg,
+			cancel,
+		)
+
+		wg.Add(1)
+		go func(cluster cluster.Cluster) {
+			defer wg.Done()
+			go func() {
+				idlerLogger.Info("Starting to watch openShift deployment configuration changes.")
+				if err := openShiftClient.WatchDeploymentConfigs(cluster.APIURL, cluster.Token, "-jenkins", controller.HandleDeploymentConfig); err != nil {
+					cancel()
+					return
+				}
+			}()
+
+			for {
+				select {
+				case <-ctx.Done():
+					idlerLogger.Infof("Stopping to watch openShift deployment configuration changes.")
+					cancel()
+					return
+				}
+			}
+		}(openShiftCluster)
+
+		wg.Add(1)
+		go func(cluster cluster.Cluster) {
```

Comment:
Just out of curiosity - why do we need to start Watch for builds and deployments in separate WGs?

Reply:
It is the same 	WaitGroup for everything. If you look at the code, you will see that I only create the WaitGroup once and then pass a pointer to it around. Basically, each go routine you start gets added to the same WaitGroup. Hence also the `wg.Add(1)`. For each cluster we do the monitoring in a separate goroutine, hence the loop.